src/Client/ConnectionString.cpp

```diff
@@ -0,0 +1,240 @@
+                                                                     : connection_string.end();
+
+    try
+    {
+        /** Poco::URI doesn't support several hosts in URI.
+          * Split string clickhouse:[user[:password]@]host1:port1, ... , hostN:portN[database]?[query_parameters]
+          * into multiple string for each host:
+          * clickhouse:[user[:password]@]host1:port1[database]?[query_parameters]
+          * ...
+          * clickhouse:[user[:password]@]hostN:portN[database]?[query_parameters]
+          */
+        Poco::URI uri;
+        const auto * last_host_begin = connection_string.begin() + offset;
+        for (const auto * it = last_host_begin; it != hosts_end; ++it)
+        {
+            if (*it == ',')
+            {
+                buildConnectionString(uri, hosts_and_ports_arguments, {last_host_begin, it}, {hosts_end, connection_string.end()});
+                last_host_begin = it + 1;
+            }
+        }
+
+        if (uri.empty())
+        {
+            // URI has no host specified
+            uri = std::string(connection_string);
+            getHostAndPort(uri, hosts_and_ports_arguments);
+        }
+        else
+            buildConnectionString(uri, hosts_and_ports_arguments, {last_host_begin, hosts_end}, {hosts_end, connection_string.end()});
+
+        Poco::URI::QueryParameters params = uri.getQueryParameters();
+        for (const auto & param : params)
+        {
+            if (param.first == "secure" || param.first == "s")
+            {
+                if (!param.second.empty())
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "secure URI query parameter does not require value");
+
+                common_arguments.push_back(makeArgument(param.first));
+            }
+            else
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "URI query parameter {} is unknown", param.first);
+        }
+
+        auto user_info = uri.getUserInfo();
+        if (!user_info.empty())
+        {
+            // Poco::URI doesn't decode user name/password by default.
+            // But ClickHouse allows to have users with email user name like: 'john@some_mail.com'
+            // john@some_mail.com should be percent-encoded: 'john%40some_mail.com'
+            std::string::size_type pos = user_info.find(':');
+            if (pos != std::string::npos)
+            {
+                common_arguments.push_back("--user");
+                common_arguments.push_back(uriDecode(user_info.substr(0, pos), true));
+
+                ++pos; // Skip ':'
+                common_arguments.push_back("--password");
+                if (user_info.size() > pos + 1)
+                    common_arguments.push_back(uriDecode(user_info.substr(pos), true));
+                else
+                {
+                    // in case of user_info == 'user:', ':' is specified, but password is empty
+                    // then add password argument "\n" which means: Ask user for a password.
```

Comment:
Interesting. Didn't know that clickhouse-client had such a shortcut.

Reply:
Small inline question about constants:
Should we add public constant `constexpr std::string_view ASK_USER_PASSWORD = "\n"sv` somewhere in ClientBase.h?
Using "\n" as a special case  for password requires adding comments in several parts of code, it is not oblivious for other non-familiar developers with code base what does "\n" means without comments.
ASK_USER_PASSWORD.

Reply:
Yes, that would be a good idea. But it is not urgent, totally fine to do this in some follow-up PR.